Easing into the 'pre-commit' framework (pcf)

safl opened the pull request:
This is a re-work of @Baekalfen initiative to introduce ``pre-commit`` at https://github.com/OpenMPDK/xNVMe/pull/48
This picks up the commits and does the following:
* Remove a handful of scripts and tools that are out-dated
* Do the clang-format using the existing clang-format definitions and adding a script using them
* Add helpers to ``Makefile`` invoking ``pre-commit`` manually
* Add ``pre-commit`` to ``.githooks/pre-commit`` with
* Add the different pre-commit hooks/plugins
* Apply them and manually fix-up lint issues in a bulk-commit
* Add the commit to git-blame-ignore
* Add a 'pre-commit' badge in README

Baekalfen commented on scripts/pcf_clang_format.py:
```diff
@@ -0,0 +1,83 @@
+    """Invoke clang-format"""
+
+    # Form arguments to clang-format parameter "-style"
+    style = []
```
Unnecessary definition.

Baekalfen commented on scripts/pcf_clang_format.py:
```diff
@@ -0,0 +1,83 @@
+
+    # Form arguments to clang-format parameter "-style"
+    style = []
+    with open(args.style_file) as sfd:
```
Unpythonic variable naming: `sfd` -> `style_file`, `f` or like that

safl replied:
Yeah, I see your point. My C abbreviation habits leaking in here. Fixing it.

safl replied:
Fixed.

Baekalfen reviewed (commented):
Tiny comments, will work regardless

Baekalfen reviewed (commented):
In commit message:

> With the option of running it on everything:
  pre-commit run clang-format --all

`--all` refers to all hooks, `--all-files` will make it run on unstaged files as well.

Baekalfen commented on Makefile:
```diff
@@ -122,6 +122,24 @@ git-setup:
+.PHONY: format-all
+format-all:
+	@echo "## xNVMe: format-all"
+	@pre-commit run --all
```
`--all` refers to all hooks, `--all-files` will make it run on unstaged files as well.

safl replied:
As discussed elsewhere, it is curious... but it is not changed to ``--all-files``.

safl replied:
Fixed.

Baekalfen commented on .githooks/pre-commit:
```diff
@@ -1,25 +1,45 @@
+You can install it via the Python Package Index:
 
-* Run the source-formatter / fix the source-format issues
+  pip3 install --user pre-commit
```
Additional step for commit message check:
```
pre-commit install --hook-type commit-msg
```

safl replied:
Doesn't ``pre-commit`` auto-install these on first-run? Along with all the others?

Baekalfen replied:
Could be it was only explicitly needed, as I was adding the commit-msg hook after the regular install.

But if we have our own pre-commit script, we need a hook in Git to trigger on commit-msg.

Baekalfen commented on .pre-commit-config.yaml:
```diff
@@ -59,3 +59,69 @@ repos:
+    args:
+    - '--max-line-length=79'
+    - '--extend-ignore=E203' # ignore "whitespace before ':'"
+- repo: https://github.com/jendrikseipp/vulture
+  rev: 'v2.3'
+  hooks:
+  - id: vulture
+    name: Python-lint-vulture
+    types: [python]
+    pass_filenames: true
+    # excluding sphinx-conf as it would break to change it
+    exclude: docs/autogen/conf.py$
```
I think we need a bit on config to make it actually trigger. But it doesn't break anything to have it in.

Baekalfen replied:
Passing filenames will presumably make `vulture` only check for dead-code within each file. You need to supply directories/packages to check inter-dependability.

Baekalfen commented on .github/workflows/verify.yml:
```diff
@@ -56,49 +56,24 @@ jobs:
-        update-alternatives --install /usr/bin/clang-format clang-format /usr/bin/clang-format-14 100
-
-    - name: xNVMe, run the source-format script
+        apt-get -qy update && apt-get -qy install shellcheck
+        pip3 install virtualenv==20.0.33
```
Shouldn't this be in `refenv/xnvme-devtools-ci:latest`?

safl replied:
Nice, catch, yes. Just added it here until it "worked". Fixing...

Baekalfen commented on .github/workflows/verify.yml:
```diff
@@ -56,49 +56,24 @@ jobs:
-      with:
-        name: source-format-issues
-        path: scripts/source-format.log
+        pre-commit run --all
```
I think you'll need `--all-files` https://github.com/pre-commit/action/blob/main/action.yml

safl replied:
Nice catch. Curiously enough it "works" with ``--all`` but that must be a fluke-accident from how the parser works... fixing...

Baekalfen commented:
Maybe I'm wrong about the `--all` vs `--all-files` arg. I can't actually find `--all` in the docs.

safl commented:
> 

I think you are right, however, for some reason ``--all`` behaves the same as ``--all-files``. Changing it to ``--all-files`` since that is what the ``--help`` says..

safl commented on .githooks/pre-commit:
```diff
@@ -1,25 +1,45 @@
+	exit 1
+}
+
+if ! pre-commit run; then
```
Hmm, yeah... this overlaps a bit with the whether or not ``GIT_HOOKS_OFF=1`` is actually needed at all.
I just kind of like that the script tells the user what to do and how to easily work around it.
But dropping the custom-script with ``GIT_HOOKS_OFF`` and relying on ``pre-commit install`` would ensure a proper invocation of ``pre-commit``. Hmm...

safl replied:
Adjusted according to offline discussion.

safl commented:
Thanks for the input @Baekalfen ! I have addressed the comments, except for one/two things:

- [ ] Question: Is the addition of ``pre-commit install --hook-type commit-msg`` required? Won't pre-commit take care of that along with along the other plugin/hook-installations?
- [ ] Whether or not ``pre-commit install`` should take of ``.githooks/pre-commit`` or whether the current approach is good enough. There is a clear argument that using ``pre-commit install`` would be preferred as it sets up ``pre-commit`` according the local environment, and that one should just use ``--no-verify`` when committing/rebasing in case one wants to ignore ``pre-commit``. Let me just mull on this one...

Baekalfen commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,135 @@
+  - id: black
+    name: Python-format-black
+    args:
+    - '--line-length=79'
```
Can't say I like the 79 character limit. That's going to be in the way

safl replied:
Bumped up to use the default provided by black.

Baekalfen commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,135 @@
+
+# Python: for all the helpers in scripts/*
+- repo: https://github.com/psf/black
+  rev: '22.1.0'
```
Should be updated to `'22.3.0'`. This fixes a bug which is blocking CI right now

Baekalfen replied:
https://github.com/psf/black/issues/2964

safl replied:
Fixed

Baekalfen commented on docs/autogen/apigen.py:
```diff
@@ -3,44 +3,48 @@
     "struct": [],
     "enum": [],
-}
+}  # type: dict
```
Is it `mypy` that requires this? Why not:
```python
DECLARATIONS : dict = {
     "func": [],
     "macro": [],
     "struct": [],
     "enum": [],
 }
```

safl replied:
Yup, mypy complains when the type-annotation is not there

Baekalfen commented on docs/autogen/apigen.py:
```diff
@@ -157,24 +160,25 @@ def main(args):
 
     pps = {}
-    for ns in syms:
-        pps[ns] = find_pp(args, syms[ns])
+    for nsprefix, val in syms.items():
+        pps[nsprefix] = find_pp(val)
```
Don't fix this unless you want to, but it's perfect for a dict-comprehension:
```python
pps = {k:find_pp(val) for k,v in syms.items()}
```

Baekalfen commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,134 @@
+    name: Python-lint-pylint
+    args:
+    - '--jobs=4' # Run with 4 threads
+    - '--good-names=i,j,k,n,f,_,Run,ex' # Add 'f' for with open()
```
`v` is also common in `for k,v in some_dict.items()`

safl replied:
Fixing...

safl replied:
Fixed.

Baekalfen reviewed (approved):
Looks good. Nothing critical

safl commented:
Thanks for the feedback. When checks pass I'll move it forward.
